pages/instructorAssessmentAccess/instructorAssessmentAccess.js

```diff
@@ -10,18 +11,164 @@ const sqlLoader = require('../../prairielib/lib/sql-loader');
+        new_rule_public.mode_raw = 'Public';
+        new_rule_public.mode = 'Public';
+        apply_rule(list, new_rule_public);
+        let new_rule_exam = Object.assign({}, new_rule);
+        new_rule_exam.mode_raw = 'Exam';
+        new_rule_exam.mode = 'Exam';
+        apply_rule(list, new_rule_exam);
+        valid = false;
+      }
+      return;
+    }
+
+    if (old_rule.valid_now_public) {
+      new_rule.valid_now_public = false;
+    }
+    if (old_rule.valid_now_exam && old_rule.exam_uuid === new_rule.exam_uuid) {
+      new_rule.valid_now_exam = false;
+    }
+
+    // Simplification: if there are multiple rules with different
+    // CBTF exam UUIDs, they are not considered a conflict, and may all
+    // be valid simultaneously.
+    if (old_rule.exam_uuid !== new_rule.exam_uuid) return;
+
+    if (
+      compare_date(new_rule.start_date_raw, old_rule.start_date_raw, -1, +1, 0) >= 0 &&
+      compare_date(new_rule.end_date_raw, old_rule.end_date_raw, +1, -1, 0) <= 0
+    ) {
+      valid = false;
+    } else if (
+      compare_date(new_rule.start_date_raw, old_rule.start_date_raw, -1, +1, 0) >= 0 &&
+      compare_date(new_rule.start_date_raw, old_rule.end_date_raw, -1, -1, -1) <= 0
+    ) {
+      new_rule.start_date_raw = old_rule.end_date_plus_one_raw;
+      new_rule.start_date = old_rule.end_date_plus_one;
+    } else if (
+      compare_date(new_rule.end_date_raw, old_rule.start_date_raw, +1, +1, +1) >= 0 &&
+      compare_date(new_rule.end_date_raw, old_rule.end_date_raw, +1, -1, 0) <= 0
+    ) {
+      new_rule.end_date_raw = old_rule.start_date_minus_one_raw;
+      new_rule.end_date = old_rule.start_date_minus_one;
+    } else if (
+      compare_date(new_rule.start_date_raw, old_rule.start_date_raw, -1, +1, 0) < 0 &&
+      compare_date(new_rule.end_date_raw, old_rule.end_date_raw, +1, -1, 0) > 0
+    ) {
+      let rule_before = Object.assign({}, new_rule);
+      rule_before.end_date_raw = old_rule.start_date_minus_one_raw;
+      rule_before.end_date = old_rule.start_date_minus_one;
+      apply_rule(list, rule_before);
+
+      new_rule.start_date_raw = old_rule.end_date_plus_one_raw;
+      new_rule.start_date = old_rule.end_date_plus_one;
+    }
+  });
+  if (valid) {
+    list.push(new_rule);
+  }
+};
+
+router.get(
+  '/',
+  asyncHandler(async (req, res, _next) => {
+    debug('GET /');
+    res.locals.assessment_settings = (
+      await sqldb.queryOneRowAsync(sql.assessment_settings, {
+        assessment_id: res.locals.assessment.id,
+      })
+    ).rows[0];
+    res.locals.access_rules = (
+      await sqldb.queryAsync(sql.assessment_access_rules, {
+        assessment_id: res.locals.assessment.id,
+        link_exam_id: config.syncExamIdAccessRules,
+      })
+    ).rows;
+
+    debug('building user-friendly description');
+    let student_rules = [];
+    let user_spec_rules = [];
+
+    // Creates sets of unique user lists
+    res.locals.access_rules.forEach((formal) => {
```

Comment:
It's not clear to me what "formal" means throughout this file.

Reply:
It refers to the original rule from the database. I renamed it for clarity.